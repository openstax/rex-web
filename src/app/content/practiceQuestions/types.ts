import { LinkedArchiveTreeSection } from '../types';

export interface State {
  summary: PracticeQuestionsSummary | null;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
  questionAnswers: QuestionAnswers;
  loading: boolean;
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

export interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export interface QuestionAnswers {
  [key: string]: PracticeAnswer | null;
}

export interface SetPracticeAnswer {
  questionId: string;
  answer: PracticeAnswer | null;
}
