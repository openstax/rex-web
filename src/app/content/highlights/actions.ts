import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
