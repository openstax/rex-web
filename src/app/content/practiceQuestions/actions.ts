import { createStandardAction } from 'typesafe-actions';
import { LinkedArchiveTreeSection } from '../types';
import { PracticeQuestion, PracticeQuestionsSummary, SetPracticeAnswer } from './types';

export const setSelectedSection = createStandardAction(
  'Content/PracticeQuestions/setSelectedSection'
)<LinkedArchiveTreeSection | null>();
export const setQuestions = createStandardAction('Content/PracticeQuestions/setQuestions')<PracticeQuestion[]>();
export const nextQuestion = createStandardAction('Content/PracticeQuestions/nextQuestion')<void>();
export const setAnswer = createStandardAction('Content/PracticeQuestions/setAnswer')<SetPracticeAnswer>();
export const finishQuestions = createStandardAction('Content/PracticeQuestions/finishQuestions')<void>();

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();

export const openPracticeQuestions = createStandardAction('Content/PracticeQuestions/open')<void>();
export const closePracticeQuestions = createStandardAction('Content/PracticeQuestions/close')<void>();
