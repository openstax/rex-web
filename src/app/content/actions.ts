import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
