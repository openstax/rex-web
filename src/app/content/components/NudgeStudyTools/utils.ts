import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
