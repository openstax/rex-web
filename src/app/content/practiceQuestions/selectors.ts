import { createSelector } from 'reselect';
import * as navigationSelectors from '../../navigation/selectors';
import * as parentSelectors from '../selectors';
import { modalUrlName } from './constants';
import { getPracticeQuestionsLocationFilters } from './utils';

export const localState = createSelector(
  parentSelectors.localState,
  (parentState) => parentState.practiceQuestions
);

export const practiceQuestionsEnabled = createSelector(
  localState,
  (state) => state.isEnabled
);

export const practiceQuestionsSummary = createSelector(
  localState,
  (state) => state.summary
);

export const hasPracticeQuestions = createSelector(
  practiceQuestionsSummary,
  (summary) => Boolean(summary && Object.keys(summary.countsPerSource).length > 0)
);

export const isPracticeQuestionsOpen = createSelector(
  hasPracticeQuestions,
  navigationSelectors.query,
  practiceQuestionsEnabled,
  (hasPQ, query, isEnabled) => hasPQ && query && query.modal === modalUrlName && isEnabled
);

export const practiceQuestionsAreLoading = createSelector(
  localState,
  (state) => state.loading
);

export const practiceQuestionsLocationFilters = createSelector(
  practiceQuestionsSummary,
  parentSelectors.book,
  getPracticeQuestionsLocationFilters
);

export const selectedSection = createSelector(
  localState,
  (state) => state.selectedSection
);

export const questionsCount = createSelector(
  localState,
  (state) => state.questions.length
);

export const currentQuestionIndex = createSelector(
  localState,
  (state) => state.currentQuestionIndex
);

export const question = createSelector(
  localState,
  currentQuestionIndex,
  (state, index) => index === null ? undefined : state.questions[index]
);

export const questionAnswers = createSelector(
  localState,
  (state) => new Map(Object.entries(state.questionAnswers))
);

export const isCurrentQuestionSubmitted = createSelector(
  questionAnswers,
  question,
  (answers, currentQuestion) => Boolean(currentQuestion && answers.has(currentQuestion.uid))
);

export const isFinalQuestion = createSelector(
  questionsCount,
  currentQuestionIndex,
  (count, index) => count - 1 === index
);

export const questionsInProggress = createSelector(
  currentQuestionIndex,
  (index) => index !== null
);

export const hasAnswers = createSelector(
  questionAnswers,
  (answers) => answers.size > 0
);
