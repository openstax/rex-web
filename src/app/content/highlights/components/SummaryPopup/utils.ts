import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
