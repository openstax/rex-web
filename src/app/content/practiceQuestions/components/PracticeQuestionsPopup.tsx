import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { push } from '../../../navigation/actions';
import * as navigation from '../../../navigation/selectors';
import { useOnEsc } from '../../../reactUtils';
import theme from '../../../theme';
import { assertDefined, assertWindow } from '../../../utils';
import Modal from '../../components/Modal';
import { bookTheme as bookThemeSelector } from '../../selectors';
import { CloseIcon, CloseIconWrapper, Header } from '../../styles/PopupStyles';
import * as pqSelectors from '../selectors';
import ShowPracticeQuestions from './ShowPracticeQuestions';

// tslint:disable-next-line: variable-name
const PracticeQuestionsPopup = () => {
  const dispatch = useDispatch();
  const popUpRef = React.useRef<HTMLElement>(null);
  const trackOpenClosePQ = useAnalyticsEvent('openClosePracticeQuestions');
  const isPracticeQuestionsOpen = useSelector(pqSelectors.isPracticeQuestionsOpen);
  const currentQuestionIndex = useSelector(pqSelectors.currentQuestionIndex);
  const bookTheme = useSelector(bookThemeSelector);
  const intl = useIntl();
  const match = useSelector(navigation.match);

  const closeAndTrack = React.useCallback((method: string) => () => {
    if (currentQuestionIndex !== null) {
      const message = intl.formatMessage({ id: 'i18n:practice-questions:popup:warning-before-close' });
      if (!assertWindow().confirm(message)) { return; }
    }

    dispatch(push(assertDefined(match, 'match should be always defined at this step')));

    trackOpenClosePQ(method);
  }, [currentQuestionIndex, trackOpenClosePQ, intl, dispatch, match]);

  useOnEsc(popUpRef, isPracticeQuestionsOpen, closeAndTrack('esc'));

  React.useEffect(() => {
    const popUp = popUpRef.current;

    if (popUp && isPracticeQuestionsOpen) {
      popUp.focus();
    }
  }, [isPracticeQuestionsOpen]);

  return isPracticeQuestionsOpen ?
    <Modal
      ref={popUpRef}
      tabIndex='-1'
      data-testid='practice-questions-popup-wrapper'
      scrollLockProps={{
        mobileOnly: false,
        onClick: closeAndTrack('overlay'),
        overlay: true,
        zIndex: theme.zIndex.highlightSummaryPopup,
      }}
    >
      <Header colorSchema={bookTheme}>
        <FormattedMessage id='i18n:practice-questions:popup:heading'>
          {(msg) => msg}
        </FormattedMessage>
        <CloseIconWrapper
          data-testid='close-practice-questions-popup'
          aria-label={intl.formatMessage({id: 'i18n:practice-questions:popup:close'})}
          data-analytics-label='Click to close Practice Questions modal'
          onClick={closeAndTrack('button')}
        >
          <CloseIcon colorSchema={bookTheme} />
        </CloseIconWrapper>
      </Header>
      <ShowPracticeQuestions />
    </Modal>
  : null;
};

export default PracticeQuestionsPopup;
