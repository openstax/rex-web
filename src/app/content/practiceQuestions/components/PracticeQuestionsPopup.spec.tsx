import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { renderToDom } from '../../../../test/reactutils';
import TestContainer from '../../../../test/TestContainer';
import { push } from '../../../navigation/actions';
import * as navigation from '../../../navigation/selectors';
import { Store } from '../../../types';
import { assertNotNull, assertWindow } from '../../../utils';
import { content } from '../../routes';
import { nextQuestion } from '../actions';
import * as pqSelectors from '../selectors';
import PracticeQuestionsPopup from './PracticeQuestionsPopup';

// this is a hack because useEffect is currently not called
// when using jsdom? https://github.com/facebook/react/issues/14050
// seems to work better in react-test-renderer but
// i need the ref here
jest.mock('react', () => {
  const react = (jest as any).requireActual('react');
  return { ...react, useEffect: react.useLayoutEffect };
});

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (children: any) => children,
}));

const mockMatch = {
  params: {
    book: { slug: 'book' },
    page: { slug: 'page' },
  },
  route: content,
  state: {},
};

describe('PracticeQuestions', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let container: HTMLElement;
  let dispatch: jest.SpyInstance;

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
    container = assertWindow().document.createElement('div');
    dispatch = jest.spyOn(store, 'dispatch');
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders practice questions modal if it is open', () => {
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>, { createNodeMock: () => container });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render practice questions modal if it is closed', () => {
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(false);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>, { createNodeMock: () => container });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('focus is on pop up content', async() => {
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    const focus = jest.fn();
    const addEventListener = jest.fn();
    const createNodeMock = () => ({focus, addEventListener});

    renderer.create(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>, {createNodeMock});

    expect(focus).toHaveBeenCalled();
  });

  it('tracks analytics and removes modal-url when clicking x icon', () => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const component = renderer.create(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>, { createNodeMock: () => container });

    renderer.act(() => {
      const closeButton = component.root.findByProps({ 'data-testid': 'close-practice-questions-popup' });
      closeButton.props.onClick();
    });

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(push(mockMatch));
  });

  it('tracks analytics and removes modal-url when clicking esc', async() => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>);

    const element = assertNotNull(node.querySelector('[data-testid=\'practice-questions-popup-wrapper\']'), '');

    element.dispatchEvent(new ((window as any).KeyboardEvent)('keydown', {key: 'Escape'}));

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(push(mockMatch));
  });

  it('tracks analytics and removes modal-url with push', async() => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>);

    const element = assertNotNull(node.querySelector('[data-testid=\'practice-questions-popup-wrapper\']'), '');

    element.dispatchEvent(new ((window as any).KeyboardEvent)('keydown', {key: 'Escape'}));

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(push(mockMatch));
  });

  it('tracks analytics and removes modal-url on overlay click', async() => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>);

    const element = assertNotNull(node.querySelector('[data-testid=\'scroll-lock-overlay\']'), '');

    const event = assertWindow().document.createEvent('MouseEvents');
    event.initEvent('click', true, true);
    const preventDefault = event.preventDefault = jest.fn();

    element.dispatchEvent(event); // this checks for bindings using addEventListener
    ReactTestUtils.Simulate.click(element, {preventDefault}); // this checks for react onClick prop

    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(push(mockMatch));
  });

  it('show warning prompt and tracks analytics after confirm', async() => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);
    const spyConfirm = jest.spyOn(assertWindow(), 'confirm')
      .mockImplementation(() => true);

    store.dispatch(nextQuestion());

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>);

    const element = assertNotNull(node.querySelector('[data-testid=\'scroll-lock-overlay\']'), '');

    const event = assertWindow().document.createEvent('MouseEvents');
    event.initEvent('click', true, true);
    const preventDefault = event.preventDefault = jest.fn();

    element.dispatchEvent(event); // this checks for bindings using addEventListener
    ReactTestUtils.Simulate.click(element, {preventDefault}); // this checks for react onClick prop

    expect(spyConfirm)
      .toHaveBeenCalledWith('Are you sure you want to exit this page? Your progress will not be saved.');
    expect(track).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(push(mockMatch));
  });

  it('show warning prompt and do not tracks analytics after cancel', async() => {
    const track = jest.spyOn(services.analytics.openClosePracticeQuestions, 'track');
    jest.spyOn(pqSelectors, 'isPracticeQuestionsOpen').mockReturnValue(true);
    jest.spyOn(navigation, 'match').mockReturnValue(mockMatch);
    track.mockClear();
    const spyConfirm = jest.spyOn(assertWindow(), 'confirm')
      .mockImplementation(() => false);

    store.dispatch(nextQuestion());

    const { node } = renderToDom(<TestContainer services={services} store={store}>
      <PracticeQuestionsPopup />
    </TestContainer>);

    const element = assertNotNull(node.querySelector('[data-testid=\'scroll-lock-overlay\']'), '');

    const event = assertWindow().document.createEvent('MouseEvents');
    event.initEvent('click', true, true);
    const preventDefault = event.preventDefault = jest.fn();

    element.dispatchEvent(event); // this checks for bindings using addEventListener
    ReactTestUtils.Simulate.click(element, {preventDefault}); // this checks for react onClick prop

    expect(spyConfirm)
      .toHaveBeenCalledWith('Are you sure you want to exit this page? Your progress will not be saved.');
    expect(track).not.toHaveBeenCalled();
    expect(dispatch).not.toHaveBeenCalledWith(push(mockMatch));
  });
});
