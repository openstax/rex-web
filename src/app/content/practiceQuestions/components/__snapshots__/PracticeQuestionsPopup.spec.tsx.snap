// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PracticeQuestions doesn't render practice questions modal if it is closed 1`] = `null`;

exports[`PracticeQuestions renders practice questions modal if it is open 1`] = `
<div
  className="PopupStyles__PopupWrapper-w0z1hl-0 cYcgUJ"
>
  <div
    className="ScrollLock__Overlay-sc-1fi845b-0 gqkuDv"
    data-testid="scroll-lock-overlay"
    onClick={[Function]}
  />
  <div
    className="PopupStyles__Modal-w0z1hl-3 ZJcjg"
    data-testid="practice-questions-popup-wrapper"
    tabIndex="-1"
  >
    <h3
      className="headings__H3-sc-1xfiggv-2 PopupStyles__Header-w0z1hl-1 kFaPzE"
    >
      Practice Questions
      <button
        aria-label="Click to close Practice Questions modal"
        className="Button__PlainButton-ayg7nk-2 PopupStyles__CloseIconWrapper-w0z1hl-4 jafVNy"
        data-analytics-label="Click to close Practice Questions modal"
        data-testid="close-practice-questions-popup"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="PopupStyles__CloseIcon-w0z1hl-5 eWfjhO"
          focusable="true"
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
    </h3>
    <div
      className="PopupStyles__PopupBody-w0z1hl-2 ShowPracticeQuestions__ShowPracticeQuestionsBody-sc-1hq8omx-0 eoLiAF"
      data-analytics-location="section not loaded"
      data-analytics-region="PQ popup"
      data-testid="show-practice-questions-body"
    >
      <div
        className="Filters-sc-1fn81fx-3 gZlQQL"
      >
        <div
          className="Filters__FiltersTopBar-sc-1fn81fx-2 fWMshY"
        >
          <div
            className="Dropdown-rmc6yw-5 fBPFfC Dropdown__TabHiddenDropDown-rmc6yw-1 kbMrSv"
          >
            <button
              aria-label="Filter PQ by Chapter & Section"
              className="Dropdown__DropdownToggle-rmc6yw-0 gpiKKF Filters__Toggle-sc-1fn81fx-1 liapzj Button__PlainButton-ayg7nk-2 chgswm"
              data-analytics-label="Filter PQ by Chapter & Section"
              onClick={[Function]}
            >
              <div
                tabIndex={-1}
              >
                Chapter & Section
                <svg
                  aria-hidden="true"
                  className="Filters__AngleIcon-sc-1fn81fx-0 jYDriK StyledIconBase-sc-bdy9j4 cCvWXa"
                  direction="down"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 320 512"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        className="ShowPracticeQuestions__ShowPracitceQuestionsContent-sc-1hq8omx-1 gWrwIb"
      >
        <div
          className="FinalScreen__StyledFinalScreen-sc-131hu2n-0 cKZyXt"
        >
          <span
            className="FinalScreen__StyledText-sc-131hu2n-1 MXQtJ"
          >
            You have reached the end of this set!
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
